### PR TITLE
missing_syscall: verify our fallback numbers when possible

### PR DESCRIPTION
**Pull request by keszybz**
Instead of defining the numbers only as fallback, always define our fallback
number, and if we have the real __NR_foo define, assert that our number matches
the real one.

This will result in warnings when our fallback number is not defined, even if
the kernel headers are new enough to define __NR_foo. This will probably annoy
people compiling for seldom-used architectures, but hopefully it'll provide
motivation to add the missing fallback defines.

The upside is that we have a higher chance of catching the cases where we got
the number wrong. Calling the wrong syscall is quite problematic, and with some
back luck, it might take us a long time to notice that we got the number wrong
on some rarely used architecture.

Also, rework some of the fallback wrappers to not call the syscall with a
negative number (that'd fail, but we'd got to the kernel and back). It seems
nicer to let the compiler know that this can never succeed.

I'm not quite convinced that this is the right thing to do, so setting "draft" for now.

### CODE REVIEW

File: src/basic/missing_syscall.h [hunk 1]
```diff
@@ -33,39 +33,47 @@ static inline int missing_pivot_root(const char *new_root, const char *put_old)
 
 /* ======================================================================= */
 
-#if !HAVE_MEMFD_CREATE
+#if defined __x86_64__
+#  define fallback_NR_memfd_create 319
```
Comment (poettering) [inline src/basic/missing_syscall.h]: is "fallback" the best name for this?

maybe use `fixed` or `known` or `validated` or so?

File: src/basic/missing_syscall.h [hunk 2]
```diff
@@ -33,39 +33,47 @@ static inline int missing_pivot_root(const char *new_root, const char *put_old)
...
+#  endif
+#  if _MIPS_SIM == _MIPS_SIM_ABI64
+#    define fallback_NR_memfd_create 5314
+#  endif
+#elif defined __i386__
+#  define fallback_NR_memfd_create 356
+#elif defined __arc__
+#  define fallback_NR_memfd_create 279
+#else
+#  warning "memfd_create number unknown for your architecture"
```
Comment (poettering) [inline src/basic/missing_syscall.h]: doesn't matter, but `memfd_create() system call numer unknown for your architecture` looks a bit better to me

**Review by poettering — APPROVED** [review]
I like this. I think this way there's a strong incentive for people to provide us with the right syscall numbers. And given the large amount of numbers you already filled in I think people can't say we are dicks breaking there stuff needlessly.

So I am all for this!

**Comment by poettering** [comment]
CI is pretty unhappy though

File: src/basic/missing_syscall.h (hunk 1 above)
Comment (keszybz) [reply src/basic/missing_syscall.h]: Not sure... what about "systemd_" ?

Comment (poettering) [reply src/basic/missing_syscall.h]: sure

**Comment by poettering** [comment]
lgtm, still thing the "fallback_" prefix is not the ideal naming

**Comment by keszybz** [comment]
bionic-amd64 failed in "udev" test:
```
TEST 30: program with subshell
device '/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda/sda5' expecting node/link 'bar9'
add:         error
test/tmpfs/dev
├── block
│   └── 8:5 -> ../sda5
├── char
├── null
└── sda5

2 directories, 3 files

remove:      ok
```
Looks unrelated.

ppc64el:
```
../src/basic/macro.h:405:9: error: static assertion failed: "__NR_memfd_create == fallback_NR_memfd_create"
         static_assert(expr, #expr)
         ^
../src/basic/missing_syscall.h:67:1: note: in expansion of macro ‘assert_cc’
 assert_cc(__NR_memfd_create == fallback_NR_memfd_create);
 ^~~~~~~~~
```
Will fix.

**Comment by keszybz** [comment]
Indeed, I pasted the number for `membarrier` instead of `memfd_create` for ppc64 :(

**Comment by keszybz** [comment]
I fixed the borked number for ppc64 and changed the prefix to "systemd_".

**Comment by keszybz** [comment]
bionic-i386 timed out in boot-and-smoke. Looks unrelated, esp. considering that i386 passed in the previous iteration. Let's merge.

**Pull request merged**